kernel/runtime/process/runtime/process.hh

```diff
@@ -0,0 +1,351 @@
+
+    /* load image */
+    auto ipc_vaddr = loadImage(pl, pm4);
+    TEST(ipc_vaddr);
+
+    /* create EC */
+    MLOG_DETAIL(mlog::app, "create EC ...", DVARhex(img.header()->entry));
+    ExecutionContext ec(capAlloc());
+    res = ec.create(kmem).as(pm4).cs(cs).sched(mythos::init::SCHEDULERS_START + 1)
+    .rawFun(reinterpret_cast<int (*)(void*)>(img.header()->entry), reinterpret_cast<void*>(*ipc_vaddr))
+    .suspended(true)
+    .invokeVia(pl).wait();
+    TEST(res);
+
+    /* create portal */
+    MLOG_DETAIL(mlog::app, "create Portal ...");
+    MLOG_DETAIL(mlog::app, "   map frame at ", DVARhex(*ipc_vaddr));
+    Frame fp(capAlloc());
+    res = fp.create(pl, kmem, 2*1024*1024, align2M).wait();
+    TEST(res);
+    res = pm4.mmap(pl, fp, *ipc_vaddr, 2*1024*1024, 0x1).wait();
+    TEST(res);
+
+    MLOG_DETAIL(mlog::app, "   create");
+    Portal port(capAlloc(), reinterpret_cast<void*>(*ipc_vaddr));
+    res = port.create(pl, kmem).wait();
+    TEST(res);
+    res = port.bind(pl, fp, 0, ec.cap()).wait();
+    TEST(res);
+
+    MLOG_DETAIL(mlog::app, "   move Portal");
+    res = myCS.move(pl, port.cap(), max_cap_depth, cs.cap(), init::PORTAL, max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(port.cap());
+
+    /* move tables */
+    MLOG_DETAIL(mlog::app, "move tables ...");
+    res = myCS.move(pl, pm3.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm3.cap());
+
+    res = myCS.move(pl, pm2.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm2.cap());
+
+    res = myCS.move(pl, pm10.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm10.cap());
+
+    res = myCS.move(pl, pm11.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm11.cap());
+
+    res = myCS.move(pl, pm12.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm12.cap());
+
+    res = myCS.move(pl, pm13.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm13.cap());
+
+    res = myCS.move(pl, pm14.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm14.cap());
+
+    res = myCS.move(pl, pm15.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm15.cap());
+
+    res = myCS.move(pl, pm16.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm16.cap());
+
+    res = myCS.move(pl, pm17.cap(), max_cap_depth, cs.cap(), pCapAlloc(), max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm17.cap());
+
+    res = myCS.move(pl, pm4.cap(), max_cap_depth, cs.cap(), init::PML4, max_cap_depth).wait();
+    TEST(res);
+    capAlloc.freeEmpty(pm4.cap());
```

Comment:
This feels like a bit too much code duplication. Can we put the caps into some kind of container (array?) and just loop?

Makes the code before that easier to read also.

Reply:
I think I will leave it like this for now.
This part of code will be obsolete and be removed when the mmap function got fixed and page tables are created automatically.